File: packages/light-source/src/audio/AudioDestination.js
/*
 * Copyright (C) 2019 Daniel Anderson
 *
 * This source code is licensed under the MIT license found in the LICENSE file in the root directory of this source tree.
 */

import { AudioDecoderType } from './AudioDecoderType.js'
import {
  AudioDestinationCapabilityFadeOut,
  AudioDestinationCapabilityPause,
  AudioDestinationCapabilityResume,
  AudioDestinationCapabilityStop,
  AudioDestinationCapabilityVolume
} from './constants.js'
import { clamp, isNumber, emptyArray } from '../util/index.js'

/**
 * An audio output buffer.
 */
export class AudioDestination {
  _native = NullNativeAudioDestination.instance // eslint-disable-line
  _decoders = emptyArray
  _rawDecoders = emptyArray

  /**
   * @returns {boolean} true if audio sources can be rendered to this destination; false if no audio can
   * be loaded or played with this destination.
   */
  isAvailable () {
    return this._native !== NullNativeAudioDestination.instance
  }

  /**
   * @returns {string[]} List of audio decoders this destination supports.
   */
  getDecoders () {
    return this._decoders
  }

  /**
   * @ignore
   */
  getRawDecoders () {
    return this._rawDecoders
  }

  /**
   * Checks if an audio decoder, defined in AudioDecoder, is supported.
   *
   * @param decoder {string} Name of decoder to check. Should be a name from AudioDecoder.
   * @returns {boolean} true if decoder is supported; otherwise, false
   */
  hasDecoder (decoder) {
    return this._rawDecoders.indexOf(decoder) >= 0
  }

  /**
   * @returns {boolean} true if the audio destination can be pause()'d; otherwise, false
   */
  canPause () {
    return this._native.hasCapability(AudioDestinationCapabilityPause)
  }

  /**
   * @returns {boolean} true if the audio destination can be resume()'d; otherwise, false
   */
  canResume () {
    return this._native.hasCapability(AudioDestinationCapabilityResume)
  }

  /**
   * @returns {boolean} true if the audio destination can be stop()'d; otherwise, false
   */
  canStop () {
    return this._native.hasCapability(AudioDestinationCapabilityStop)
  }

  /**
   * @returns {boolean} true if the audio destination can be faded out; otherwise, false
   */
  canFadeOut () {
    return this._native.hasCapability(AudioDestinationCapabilityFadeOut)
  }

  /**
   * @returns {boolean} true if the audio destination has volume controls; otherwise, false
   */
  hasVolume () {
    return this._native.hasCapability(AudioDestinationCapabilityVolume)
  }

  /**
   * @returns volume {number} Current value [0-1] of the volume of the audio destination.
   */
  getVolume () {
    return this._native.volume
  }

  /**
   * Sets the volume.
   *
   * @param value {number} A value between [0-1]. If the value is not a number, volume is set to 0. If the value is
   * out of range, it is Math.clamp()'d to [0-1].
   */
  setVolume (value) {
    this._native.volume = isNumber(value) ? clamp(value, 0, 1) : 0
  }

  /**
   * Pause playback of the output buffer.
   */
  pause () {
    this._native.pause()
  }

  /**
   * Resume playback of the output buffer.
   */
  resume () {
    this._native.resume()
  }

  /**
   * Stop playback of the audio buffer.
   *
   * @param {Number} [opts.fadeOutMs] Time in milliseconds to fade out the audio before stopping. Only available if
   * canFadeOut is true.
   */
  stop (opts) {
    this._native.stop(opts)
  }

  /**
   * @ignore
   */
  $createNativeAudioSource () {
    return this._native.createAudioSource()
  }

  /**
   * @ignore
   */
  $setNative (nativeDestination) {
    this._native = nativeDestination || NullNativeAudioDestination.instance
    this._decoders = []

    for (const rawDecoder of this._native.decoders) {
      if (rawDecoder in AudioDecoderType) {
        this._decoders.push(rawDecoder)
      }
    }

    this._rawDecoders = [...this._native.decoders]

    Object.freeze(this._rawDecoders)
    Object.freeze(this._decoders)
  }
}

/**
 * @ignore
 */
class NullNativeAudioDestination {
  static instance = new NullNativeAudioDestination()
  decoders = emptyArray
  resume () {}
  setVolume (value) {}
  getVolume () { return 0 }
  stop () {}
  pause () {}
  hasCapability () { return false }
}

File: packages/@lse/core/addon/lse-lib-audio/lse/AudioDestination.cc
/*
 * Copyright (C) 2020 Daniel Anderson
 *
 * This source code is licensed under the MIT license found in the LICENSE file in the root directory of this source
 * tree.
 */

#include "AudioDestination.h"

#define CHECK_IMPL(impl) if (!(impl)) { throw Napi::Error::New(info.Env(), "AudioDestination is not initialized!"); }

namespace lse {

AudioDestination::AudioDestination(const Napi::CallbackInfo& info) : Napi::SafeObjectWrap<AudioDestination>(info) {
}

AudioDestination::~AudioDestination() {
  if (this->impl) {
    this->impl->Finalize();
  }
}

void AudioDestination::Constructor(const Napi::CallbackInfo& info) {
  this->impl = Napi::ConstructorWithExternalFactory<AudioDestinationInterface, AudioDestinationInterfaceFactory>(
      info, "AudioDestination");
}

Napi::Function AudioDestination::GetClass(Napi::Env env) {
  static Napi::FunctionReference constructor;

  if (constructor.IsEmpty()) {
    Napi::HandleScope scope(env);

    constructor = AudioDestination::DefineClass(env, "AudioDestination", true, {
        AudioDestination::InstanceMethod("createAudioSource", &AudioDestination::CreateAudioSource),
        AudioDestination::InstanceMethod("resume", &AudioDestination::Resume),
        AudioDestination::InstanceMethod("pause", &AudioDestination::Pause),
        AudioDestination::InstanceMethod("stop", &AudioDestination::Stop),
        AudioDestination::InstanceMethod("destroy", &AudioDestination::Destroy),
        AudioDestination::InstanceMethod("hasCapability", &AudioDestination::HasCapability),
        AudioDestination::InstanceAccessor("volume", &AudioDestination::GetVolume, &AudioDestination::SetVolume),
        AudioDestination::InstanceAccessor("decoders", &AudioDestination::GetDecoders),
    });
  }

  return constructor.Value();
}

void AudioDestination::Destroy(const Napi::CallbackInfo& info) {
  if (this->impl) {
    this->impl->Destroy(info);
    this->impl->Finalize();
    this->impl = nullptr;
  }
}

Napi::Value AudioDestination::GetDecoders(const Napi::CallbackInfo& info) {
  CHECK_IMPL(this->impl);
  return this->impl->GetDecoders(info);
}

Napi::Value AudioDestination::CreateAudioSource(const Napi::CallbackInfo& info) {
  CHECK_IMPL(this->impl);
  return this->impl->CreateAudioSource(info);
}

void AudioDestination::Resume(const Napi::CallbackInfo& info) {
  CHECK_IMPL(this->impl);
  this->impl->Resume(info);
}

void AudioDestination::Pause(const Napi::CallbackInfo& info) {
  CHECK_IMPL(this->impl);
  this->impl->Pause(info);
}

void AudioDestination::Stop(const Napi::CallbackInfo& info) {
  CHECK_IMPL(this->impl);
  this->impl->Stop(info);
}

Napi::Value AudioDestination::GetVolume(const Napi::CallbackInfo& info) {
  CHECK_IMPL(this->impl);
  return this->impl->GetVolume(info);
}

void AudioDestination::SetVolume(const Napi::CallbackInfo& info, const Napi::Value& value) {
  CHECK_IMPL(this->impl);
  this->impl->SetVolume(info, value);
}

Napi::Value AudioDestination::HasCapability(const Napi::CallbackInfo& info) {
  CHECK_IMPL(this->impl);
  return this->impl->HasCapability(info);
}

void AudioDestination::Finalize() {
  throw std::runtime_error("Not implemented");
}

} // namespace lse

#undef CHECK_IMPL
